src/pybind/mgr/dashboard/services/access_control.py

```diff
@@ -348,363 +350,297 @@ def load(cls):
-               'name=permissions,type=CephString,n=N',
-        'desc': 'Add the scope permissions for a role',
-        'perm': 'w'
-    },
-    {
-        'cmd': 'dashboard ac-role-del-scope-perms '
-               'name=rolename,type=CephString '
-               'name=scopename,type=CephString',
-        'desc': 'Delete the scope permissions for a role',
-        'perm': 'w'
-    },
-    {
-        'cmd': 'dashboard ac-user-show '
-               'name=username,type=CephString,req=false',
-        'desc': 'Show user info',
-        'perm': 'r'
-    },
-    {
-        'cmd': 'dashboard ac-user-create '
-               'name=username,type=CephString '
-               'name=password,type=CephString,req=false '
-               'name=rolename,type=CephString,req=false '
-               'name=name,type=CephString,req=false '
-               'name=email,type=CephString,req=false',
-        'desc': 'Create a user',
-        'perm': 'w'
-    },
-    {
-        'cmd': 'dashboard ac-user-delete '
-               'name=username,type=CephString',
-        'desc': 'Delete user',
-        'perm': 'w'
-    },
-    {
-        'cmd': 'dashboard ac-user-set-roles '
-               'name=username,type=CephString '
-               'name=roles,type=CephString,n=N',
-        'desc': 'Set user roles',
-        'perm': 'w'
-    },
-    {
-        'cmd': 'dashboard ac-user-add-roles '
-               'name=username,type=CephString '
-               'name=roles,type=CephString,n=N',
-        'desc': 'Add roles to user',
-        'perm': 'w'
-    },
-    {
-        'cmd': 'dashboard ac-user-del-roles '
-               'name=username,type=CephString '
-               'name=roles,type=CephString,n=N',
-        'desc': 'Delete roles from user',
-        'perm': 'w'
-    },
-    {
-        'cmd': 'dashboard ac-user-set-password '
-               'name=username,type=CephString '
-               'name=password,type=CephString',
-        'desc': 'Set user password',
-        'perm': 'w'
-    },
-    {
-        'cmd': 'dashboard ac-user-set-info '
-               'name=username,type=CephString '
-               'name=name,type=CephString '
-               'name=email,type=CephString',
-        'desc': 'Set user info',
-        'perm': 'w'
-    }
-]
-
-
-def handle_access_control_command(cmd):
-    if cmd['prefix'] == 'dashboard set-login-credentials':
-        username = cmd['username']
-        password = cmd['password']
-        try:
-            user = ACCESS_CTRL_DB.get_user(username)
-            user.set_password(password)
-        except UserDoesNotExist:
-            user = ACCESS_CTRL_DB.create_user(username, password, None, None)
-            user.set_roles([ADMIN_ROLE])
-
-        ACCESS_CTRL_DB.save()
 
-        return 0, '''\
+@CLIWriteCommand('dashboard set-login-credentials',
+                 'name=username,type=CephString '
+                 'name=password,type=CephString',
+                 'Set the login credentials')
```

Comment:
It's a pity that Python 2.7 does not support type hints (https://www.python.org/dev/peps/pep-0484/). It'd be really nice to define the Ceph cmd prototype from the type hints.

A few extra-magic suggestions:
- Avoid repeating the `dashboard` part. Maybe creating a top level constant `mgr.MODULE_NAME`?
- Create the command name from the `fun.func_name.replace("_", "-")`?
- Taking the command description from the docstring `fun.fun_doc`?

That would result in something like:
```python
@CLIWriteCommand('name=username,type=CephString name=password,type=CephString')
def set_login_credentials_cmd(_, username, password):
    """Set the login credentials"""
```

Reply:
I thought about all these things when I was implementing this decorator approach.

> Avoid repeating the dashboard part. Maybe creating a top level constant mgr.MODULE_NAME?

Not all modules use the name of the module as the prefix of the command, and therefore I decided not to do any automation.

> Create the command name from the fun.func_name.replace("_", "-")?

Some modules like to separate command words with spaces, others with `-`. Therefore I decided to keep the command specification explicit.

> Taking the command description from the docstring fun.fun_doc?

The reason for this is because time is limited and I still need to implement other features in the dashboard before the feature freeze. IMO, the current implementation is a good balance between the usefulness/usability of the approach and the time it took to implement.

Reply:
> Taking the command description from the docstring fun.fun_doc?

This would suddenly make documentation part of the code, which is unexpected in my opinion. Instead, I'd favor a DSL to generate these cli command descriptions.

Reply:
> This would suddenly make documentation part of the code
Well, technically docstrings are code rather than merely docs. That's the way `doctest` works, for example.